src/unix/socket.c

```diff
@@ -571,21 +631,36 @@ static sysreturn unixsock_accept4(struct sock *sock, struct sockaddr *addr,
 sysreturn unixsock_sendto(struct sock *sock, void *buf, u64 len, int flags,
         struct sockaddr *dest_addr, socklen_t addrlen)
 {
-    /* Non-connected sockets are not supported, so destination address is
-     * ignored. */
-    return apply(sock->f.write, buf, len, 0, current, false,
-            syscall_io_complete);
+    unixsock s = (unixsock) sock;
+    if (dest_addr || addrlen) {
+        if ((sock->type & SOCK_STREAM)) {
+            if (s->peer)
+                return -EISCONN;
+            else
+                return -ENOTCONN;
+        }
+        if (!(dest_addr && addrlen))
+            return -EFAULT;
+        if (addrlen < sizeof(struct sockaddr_un))
+            return -EINVAL;
+        if (dest_addr && !validate_user_memory(dest_addr, addrlen, true))
+            return -EFAULT;
+    }
+    return unixsock_write_with_addr(s, buf, len, 0, current, false, syscall_io_complete, (struct sockaddr_un *)dest_addr, addrlen);
 }
 
 sysreturn unixsock_recvfrom(struct sock *sock, void *buf, u64 len, int flags,
-        struct sockaddr *dest_addr, socklen_t *addrlen)
+        struct sockaddr *src_addr, socklen_t *addrlen)
 {
-    /* Non-connected sockets are not supported, so source address is not set. */
-    if (addrlen) {
-        *addrlen = 0;
+    if (src_addr || addrlen) {
+        if (!(src_addr && addrlen))
+            return -EFAULT;
+        if (!validate_user_memory(addrlen, sizeof(socklen_t), true) ||
+            !validate_user_memory(src_addr, *addrlen, true))
```

Comment:
These validate_user_memory() calls are already in the generic recvfrom() function in net/netsyscall.c, so they shouldn't be necessary here.

Reply:
Fixed